### PR TITLE
EC-896 Save comments for code users

### PR DESCRIPTION
**Pull request by John-Williams**
https://nicedigital.atlassian.net/browse/EC-896

### CODE REVIEW

**Review by suegarner — CHANGES_REQUESTED** [review]

File: Comments/Services/AnswerService.cs [hunk 1]
```diff
@@ -61,31 +71,58 @@ public AnswerService(ConsultationsContext consultationsContext, IUserService use
...
         {
-            if (!_currentUser.IsAuthorised)
-                return (answer: null, validate: new Validate(valid: false, unauthorised: true, message: "Not logged in creating answer"));
-
-	        var status = _context.GetStatus(StatusName.Draft);
-	        var question = _context.GetQuestion(answer.QuestionId);
-
-            var answerToSave = new Models.Answer(answer.QuestionId, _currentUser.UserId, answer.AnswerText, answer.AnswerBoolean, question, status.StatusId, null);
+            if (!_currentUser.IsAuthenticated)
+                return (answer: null, validate: new Validate(valid: false, unauthenticated: true, message: "Not logged in creating answer"));
```
Comment (suegarner) [inline Comments/Services/AnswerService.cs:102]: Is the answer message displayed to the user? Could do with rewriting the message

**Review by suegarner — CHANGES_REQUESTED** [review]

File: Comments.Test/Infrastructure/FakeHttpContextAccessor.cs [hunk 2]
```diff
@@ -28,14 +30,29 @@ public static IHttpContextAccessor Get(bool isAuthenticated, string displayName
 				var claims = new List<Claim>
                 {
                     new Claim(ClaimType.DisplayName, displayName, null, AuthenticationConstants.IdAMIssuer),
-                    new Claim(ClaimType.NameIdentifier, userId.ToString(), null, AuthenticationConstants.IdAMIssuer),
-					new Claim(ClaimType.Organisations, JsonConvert.SerializeObject(new List<Organisation>(){ new Organisation(1, "NICE", true)}), "www.nice.org.uk", AuthenticationConstants.IdAMIssuer)
+                    //new Claim(ClaimType.Organisations, JsonConvert.SerializeObject(new List<Organisation>(){ new Organisation(1, "NICE", true)}), "www.nice.org.uk", AuthenticationConstants.IdAMIssuer)
```
[the remark this answers is not in the document]
Comment (John-Williams) [reply Comments.Test/Infrastructure/FakeHttpContextAccessor.cs:33]: done

File: Comments.Test/IntegrationTests/API/Answers/DeleteAnswerUsingOrganisationSessionCookieTests.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+			context.Database.EnsureDeleted();
+
+			var sourceURI = $"consultations://./consultation/{_consultationId}/document/1/chapter/introduction";
+
+			var organisationAuthorisationId = TestBaseDBHelpers.AddOrganisationAuthorisationWithLocation(1, _consultationId, context, null, "123412341234");
+			TestBaseDBHelpers.AddOrganisationUser(context, organisationAuthorisationId, _sessionId, null, organisationUserId);
+
+			var locationId = TestBaseDBHelpers.AddLocation(context, sourceURI);
+			var _questionId = TestBaseDBHelpers.AddQuestion(context, locationId);
+			var _existingAnswerId = TestBaseDBHelpers.AddAnswer(context, _questionId, statusId: StatusId, organisationUserId: organisationUserId);
```
[the remark this answers is not in the document]
Comment (John-Williams) [reply Comments.Test/IntegrationTests/API/Answers/DeleteAnswerUsingOrganisationSessionCookieTests.cs:40]: updated

File: Comments.Test/IntegrationTests/API/Answers/DeleteAnswerUsingOrganisationSessionCookieTests.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+using Shouldly;
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Comments.Test.IntegrationTests.API.Comments
```
[the remark this answers is not in the document]
Comment (John-Williams) [reply Comments.Test/IntegrationTests/API/Answers/DeleteAnswerUsingOrganisationSessionCookieTests.cs:16]: updated

File: Comments.Test/IntegrationTests/API/Answers/EditAnswerUsingOrganisationSessionCookieTests.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	public class EditAnswerUsingOrganisationSessionCookieTests : TestBaseLight
+	{
+		private static readonly Guid _sessionId = Guid.Parse("11111111-1111-1111-1111-111111111111");
+
+		private readonly TestServer _server;
+		private readonly HttpClient _client;
+		private readonly int _consultationId = 1;
+		private readonly int _questionId;
+		private readonly int _existingAnswerId;
+		const int StatusId = 1;
```
[the remark this answers is not in the document]
Comment (John-Williams) [reply Comments.Test/IntegrationTests/API/Answers/EditAnswerUsingOrganisationSessionCookieTests.cs:28]: updated

File: Comments.Test/IntegrationTests/API/Comments/CommentTests.cs [hunk 6]
```diff
@@ -198,6 +201,9 @@ public async Task Edit_Comment()
             result.comment.CommentText.ShouldBe(updatedCommentText);
         }
 
-	    
+
+        
+
+
```
[the remark this answers is not in the document]
Comment (John-Williams) [reply Comments.Test/IntegrationTests/API/Comments/CommentTests.cs:207]: deleted

File: Comments.Test/IntegrationTests/API/Comments/DeleteCommentUsingOrganisationSessionCookieTests.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+namespace Comments.Test.IntegrationTests.API.Comments
+{
+	public class DeleteCommentUsingOrganisationSessionCookieTests : TestBaseLight
+	{
+		private static readonly Guid _sessionId = Guid.Parse("11111111-1111-1111-1111-111111111111");
+
+		private readonly TestServer _server;
+		private readonly HttpClient _client;
+		private readonly int consultationId = 1;
+		private readonly Comment comment;
```
[the remark this answers is not in the document]
Comment (John-Williams) [reply Comments.Test/IntegrationTests/API/Comments/DeleteCommentUsingOrganisationSessionCookieTests.cs:25]: updated

File: Comments/Services/AnswerService.cs (hunk 1 above)
Comment (John-Williams) [reply Comments/Services/AnswerService.cs:102]: this error message should never be hit by the front-end. if you're not logged in, it should show you the login banner component.
the only way this error would happen is if the front-end had a bug in it, or a user used postman/curl etc to hit the api, without an authentication cookie.

**Review by suegarner — APPROVED** [review]

**Pull request merged**